TFG-Labs/custom-shipping-step-by-items
# [WEB-438] -Autocomplete improvement

### What problem is this solving?
This change is meant to encourage users to enter addresses using google autocomplete as well as . 

Desired change is not achieved exactly as required
Reason: You cannot access/manipulate returned google autocomplete addresses data until the users click on a suggested address (i.e through debouncing).

What has been achieved?

- **requests** for autocomplete addresses only executes **after 2 input value chars** are inserted by the user - this reduces requests
- if user backspaces to **less than 3 input chars**, google **autocomplete** and it's listener is **removed** to reduce requests as it stays mounted otherwise.
- To reduce premature error messaging for addresses  :
           1. A function **isSlowerNetwork** checks user network speed, if user is on slower network - **delay to error messages** is increased to 3s - from testing with throttled network this autocomplete response will not always be recieved in 3seconds.
           2. Before **checkAddressResults** function is **called** to evaluate if response has been received in order to display not found address message - I've put in logic to prevent checkAddressResults from executing if user is **still typing** **number**(house number) **or** has just put in a **space** in order to start typing street name. This will decrease premature address error message

### Used delay times stipulated in ticket for 'faster' networks and 3s for slow networks - behaviour looks like it needs longer delay times

#### How it works
Detailed above

#### How to test it?
- Test a few a few addresses and use checklist of what's been achieved above against current behaviour in prod.
- Test with fast and slow network.
- Check if code changes are efficient.

[Workspace](https://autocompleteweb436--thefoschini.myvtex.com/checkout/#/shipping)

### Screenshots/Video example usage:
- Desktop
- Tablet
- Movil
### Faster speed
https://github.com/user-attachments/assets/9cc648ba-d5ba-4ee1-a545-5f86b2fee2eb

### Throttled netwok
https://github.com/user-attachments/assets/62a9f9ee-78e7-4dcc-a0a7-0ba6681919c7




### Related to / Depends on

#### How does this PR make you feel? :link:
![]()
